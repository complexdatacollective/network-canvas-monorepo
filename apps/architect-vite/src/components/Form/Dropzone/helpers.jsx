const getExtension = (path) => {
  const match = /(.[A-Za-z0-9]+)$/.exec(path);
  if (!match) { return null; }
  return match[1];
};

const matchExtension = (path, extension) => RegExp(`${extension}$`).test(path.toLowerCase());

const acceptsPath = (accepts) => (path) => accepts.some((accept) => matchExtension(path, accept));

export const acceptsPaths = (accepts, paths) => {
  if (!paths || paths.length === 0) { return false; }
  return paths.every(acceptsPath(accepts));
};

export const getRejectedExtensions = (accepts, paths) => paths.reduce((memo, path) => {
  if (acceptsPath(accepts)(path)) { return memo; }
  const extension = getExtension(path);
  if (memo.includes(extension)) { return memo; }
  return [...memo, extension];
}, []);

export const getAcceptsExtensions = (accepts) => accepts.map((accept) => accept.substr(1));
