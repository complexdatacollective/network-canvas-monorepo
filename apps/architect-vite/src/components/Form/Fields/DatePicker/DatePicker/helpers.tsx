import { DateTime, Info } from "luxon";
import { isEqual, get } from "lodash";

export const now = () => DateTime.local();

export interface DateObject {
	year: number | null;
	month: number | null;
	day: number | null;
}

/**
 * Is date object fully complete?
 */
export const isComplete =
	(type: string | null) =>
	({ day, month, year }: DateObject) => {
		switch (type) {
			case "year":
				return !!year;
			case "month":
				return !!year && !!month;
			default:
				return !!year && !!month && !!day;
		}
	};

/**
 * Is date object empty
 */
export const isEmpty = () => (date: DateObject) => isEqual(date, { year: null, month: null, day: null });

export const getMonthName = (numericMonth: number) => get(Info.months(), numericMonth - 1, numericMonth);

export interface RangeItem {
	value: number;
	label: string | number;
	isOutOfRange?: boolean;
}

export const formatRangeItem = (value: number, props: Partial<RangeItem> = {}): RangeItem => ({
	value,
	label: value,
	...props,
});
