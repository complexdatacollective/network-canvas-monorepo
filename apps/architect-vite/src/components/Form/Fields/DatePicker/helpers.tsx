import { DateTime, Info } from "luxon";
import { difference, intersection, get } from "lodash";

export const now = () => DateTime.local();

export const isEmpty = (value: unknown) => value === null || value === "";

export const getFirstDayOfMonth = (dateObj: { year?: number | null; month?: number | null; day?: number | null }) =>
	DateTime.fromObject({ ...dateObj, day: 1 }).toFormat("c");

export const asNullObject = (keys: string[]) => keys.reduce((acc, key) => ({ ...acc, [key]: null }), {});

export const getProperties = (obj: Record<string, any>) =>
	Object.keys(obj).reduce<string[]>((acc, key) => {
		if (!obj[key]) {
			return acc;
		}
		return [...acc, key];
	}, []);

export const hasProperties =
	(includes: string[] = [], excludes: string[] = []) =>
	(obj: Record<string, any>) => {
		const props = getProperties(obj);
		const hasIncludes = difference(includes, props).length === 0;
		const noExcludes = intersection(props, excludes).length === 0;
		return hasIncludes && noExcludes;
	};

export const getMonthName = (numericMonth: number) => get(Info.months(), numericMonth - 1, numericMonth);
