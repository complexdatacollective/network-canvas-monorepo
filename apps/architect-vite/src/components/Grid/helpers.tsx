export const convertSize = (size: number): string => {
	switch (size) {
		case 2:
			return "MEDIUM";
		case 4:
			return "LARGE";
		default:
			return "SMALL";
	}
};

export const parseSize = (size: string): number => {
	switch (size) {
		case "MEDIUM":
			return 2;
		case "LARGE":
			return 4;
		default:
			return 1;
	}
};

type GridItem = {
	size: string;
	id: string;
};

export const getRemainingSpace = (items: GridItem[] = [], capacity = 0): number =>
	items.reduce((acc, { size }) => acc - parseSize(size), capacity);

export const trimSize = (from: number, to: number, items: GridItem[], capacity: number): number => {
	const remainingSpace = getRemainingSpace(items, capacity);
	if (to !== 3) {
		return to;
	}
	if (from < 3 && remainingSpace === 4) {
		return 4;
	}
	return 2;
};

export const getLayout = (items: GridItem[] = [], capacity = 4) => {
	const remainingSpace = getRemainingSpace(items, capacity);

	const layout = items.reduce(
		(memo, { id, size }) => {
			const y = memo.reduce((acc, { h }) => acc + h, 0);
			const h = parseSize(size);
			const maxH = h + remainingSpace === 3 ? 2 : h + remainingSpace;

			return [
				...memo,
				{
					i: id,
					y,
					w: 1,
					h,
					x: 0,
					maxH,
				},
			];
		},
		[
			{
				/**
				 * Fixes a bug with react-grid-layout not updating when layout prop
				 * is updated, but hasn't changed, react-grid-layout may ignore prop.
				 * This forces an update every time.
				 */
				i: Math.random().toString(),
				y: 0,
				w: 0,
				h: 0,
				x: 0,
			},
		],
	);

	return layout;
};
