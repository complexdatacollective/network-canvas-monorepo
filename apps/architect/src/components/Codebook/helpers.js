import {
  reduce, get, compact, uniq, map,
} from 'lodash';
import { getType, getAllVariablesByUUID } from '@selectors/codebook';
import { makeGetIsUsed } from '@selectors/codebook/isUsed';
import { getVariableIndex } from '@selectors/indexes';
import { getProtocol, getCodebook } from '@selectors/protocol';

const getIsUsed = makeGetIsUsed({ formNames: [] });

/**
 * Extract basic stage meta by index from the app state
 * @param {Object} state Application state
 * @returns {Object[]} Stage meta sorted by index in state
 */
export const getStageMetaByIndex = (state) => {
  const protocol = getProtocol(state);
  return protocol.stages
    .map(({ label, id }) => ({ label, id }));
};

export const getVariableMetaByIndex = (state) => {
  const codebook = getCodebook(state);
  const variables = getAllVariablesByUUID(codebook);
  return variables;
};

/**
 * Extract the stage name from a path string
 * @param {string} path {}
 * @returns {string | null} return a stageIndex or null if stage not found.
 */
const getStageIndexFromPath = (path) => {
  const matches = /stages\[([0-9]+)\]/.exec(path);
  return get(matches, 1, null);
};

const codebookVariableReferenceRegex = /codebook\.(ego|node\[([^\]]+)\]|edge\[([^\]]+)\])\.variables\[(.*?)\].validation\.(sameAs|differentFrom)/;

export const getCodebookVariableIndexFromValidationPath = (path) => {
  const match = path.match(codebookVariableReferenceRegex);

  return get(match, 4, null);
};

/**
 * Takes an object in the format of `{[path]: variableID}` and a variableID to
 * search for. Returns an array of paths that match the variableID.
 *
 * @param {Object.<string, string>}} index Usage index in (in format `{[path]: variableID}`)
 * @param {any} value Value to match in usage index
 * @returns {string[]} List of paths ("usage array")
 */
export const getUsage = (index, value) => reduce(index, (acc, indexValue, path) => {
  if (indexValue !== value) { return acc; }
  return [...acc, path];
}, []);

/**
 * Get stage meta (wtf is stage meta, Steve? ðŸ¤¦) that matches "usage array"
 * (with duplicates removed).
 *
 * See `getUsage()` for how the usage array is generated.
 *
 * Any stages that can't be found in the index are omitted.
 *
 * @param {Object[]} stageMetaByIndex Stage meta by index (as created by `getStageMetaByIndex()`)
 * @param {Object[]} variableMetaByIndex Variable meta by index (as created by
 * `getVariableMetaByIndex()`)
 * @param {string[]} usageArray "Usage array" as created by `getUsage()`
 * @returns {Object[]} List of stage meta `{ label, id }`.
 */
export const getUsageAsStageMeta = (stageMetaByIndex, variableMetaByIndex, usageArray) => {
  // Filter codebook variables from usage array
  const codebookVariablePaths = usageArray.filter(getCodebookVariableIndexFromValidationPath);
  const codebookVariablesWithMeta = codebookVariablePaths.map((path) => {
    const variableId = getCodebookVariableIndexFromValidationPath(path);
    const { name } = variableMetaByIndex[variableId];
    return {
      label: `Used as validation for "${name || 'unknown'}"`,
    };
  });

  const stageIndexes = compact(uniq(usageArray.map(getStageIndexFromPath)));
  const stageVariablesWithMeta = stageIndexes.map(
    (stageIndex) => get(stageMetaByIndex, stageIndex),
  );

  return [
    ...stageVariablesWithMeta,
    ...codebookVariablesWithMeta,
  ];
};

/**
 * Helper function to be used with Array.sort. Sorts a collection of variable
 * definitions by the label property.
 *
 * @param {Object} a { label: string }
 * @param {Object} b { label: string }
 * @returns {number} -1 if a < b, 1 if a > b, 0 if a === b
 */
export const sortByLabel = (a, b) => {
  if (a.label < b.label) { return -1; }
  if (a.label > b.label) { return 1; }
  return 0;
};

/**
 * Returns entity meta data for use in the codebook.
 * @param {*} state
 * @param {*} param1
 * @returns
 */
export const getEntityProperties = (state, { entity, type }) => {
  const {
    name,
    color,
    variables,
  } = getType(state, { entity, type });

  const variableIndex = getVariableIndex(state);
  const variableMeta = getVariableMetaByIndex(state);
  const stageMetaByIndex = getStageMetaByIndex(state);
  const isUsedIndex = getIsUsed(state);

  const variablesWithUsage = map(
    variables,
    (variable, id) => {
      const inUse = get(isUsedIndex, id, false);

      const baseProperties = {
        ...variable,
        id,
        inUse,
      };

      if (!inUse) {
        return (baseProperties);
      }

      const usage = getUsageAsStageMeta(
        stageMetaByIndex,
        variableMeta,
        getUsage(variableIndex, id),
      ).sort(sortByLabel);

      const usageString = usage.map(({ label }) => label).join(', ').toUpperCase();
      return ({
        ...baseProperties,
        usage,
        usageString,
      });
    },
  );

  return {
    name,
    color,
    variables: variablesWithUsage,
  };
};
