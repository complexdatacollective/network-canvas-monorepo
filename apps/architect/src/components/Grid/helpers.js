export const convertSize = (size) => {
  switch (size) {
    case 2:
      return 'MEDIUM';
    case 4:
      return 'LARGE';
    default:
      return 'SMALL';
  }
};

export const parseSize = (size) => {
  switch (size) {
    case 'MEDIUM':
      return 2;
    case 'LARGE':
      return 4;
    default:
      return 1;
  }
};

export const getRemainingSpace = (items = [], capacity = 0) => items.reduce(
  (acc, { size }) => acc - parseSize(size),
  capacity,
);

export const trimSize = (from, to, items, capacity) => {
  const remainingSpace = getRemainingSpace(items, capacity);
  if (to !== 3) { return to; }
  if (from < 3 && remainingSpace === 4) { return 4; }
  return 2;
};

export const getLayout = (items = [], capacity = 4) => {
  const remainingSpace = getRemainingSpace(items, capacity);

  const layout = items.reduce(
    (memo, { id, size }) => {
      const y = memo.reduce((acc, { h }) => acc + h, 0);
      const h = parseSize(size);
      const maxH = h + remainingSpace === 3 ? 2 : h + remainingSpace;

      return [
        ...memo,
        {
          i: id,
          y,
          w: 1,
          h,
          x: 0,
          maxH,
        },
      ];
    },
    [{
      /**
       * Fixes a bug with react-grid-layout not updating when layout prop
       * is updated, but hasn't changed, react-grid-layout may ignore prop.
       * This forces an update every time.
       */
      i: Math.random().toString(),
      y: 0,
      w: 0,
      h: 0,
      x: 0,
    }],
  );

  return layout;
};
