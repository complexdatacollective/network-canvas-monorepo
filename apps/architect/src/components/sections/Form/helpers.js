import { formValueSelector } from 'redux-form';
import { omit, get, reduce } from 'lodash';
import { getVariablesForSubject } from '@selectors/codebook';

export const CODEBOOK_PROPERTIES = ['options', 'parameters', 'component', 'validation'];

export const getCodebookProperties = (properties) => reduce(
  CODEBOOK_PROPERTIES,
  (memo, key) => {
    const property = properties[key];
    if (!Object.keys(properties).includes(key)) { return memo; }
    return {
      ...memo,
      [key]: property,
    };
  },
  {},
);

export const normalizeField = (field) => omit(field, ['id', ...CODEBOOK_PROPERTIES]);

// Merge item with variable info from codebook
export const itemSelector = (entity, type) => (state, { form, editField }) => {
  const item = formValueSelector(form)(state, editField);

  if (!item) { return null; }

  const variable = item && item.variable;

  const codebookVariables = getVariablesForSubject(state, { entity, type });
  const codebookVariable = get(codebookVariables, variable, {});
  const codebookProperties = getCodebookProperties(codebookVariable);

  return {
    ...item,
    ...codebookProperties,
  };
};
