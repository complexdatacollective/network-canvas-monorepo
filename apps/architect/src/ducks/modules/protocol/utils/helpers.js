/* eslint-disable import/prefer-default-export */

import { size, get } from 'lodash';
import { COLOR_PALETTE_BY_ENTITY, COLOR_PALETTES } from '../../../../config';

export const getNextCategoryColor = (protocol, entity) => {
  if (!protocol || !entity) { return null; }
  const paletteName = entity === 'edge'
    ? COLOR_PALETTE_BY_ENTITY.edge
    : COLOR_PALETTE_BY_ENTITY.node;
  const paletteSize = COLOR_PALETTES[paletteName];
  const typeCount = size(get(protocol, ['codebook', entity], {}));
  const nextNumber = (typeCount % paletteSize) + 1;
  const nextColor = `${paletteName}-${nextNumber}`;

  return nextColor;
};
