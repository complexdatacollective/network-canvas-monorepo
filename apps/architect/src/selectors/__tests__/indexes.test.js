/* eslint-env jest */

import { getMockState } from '@app/__tests__/helpers';

import {
  getVariableIndex,
  getAssetIndex,
  getNodeIndex,
  getEdgeIndex,
  utils,
} from '../indexes';

const testState = getMockState();

describe('indexes selectors', () => {
  describe('utils.buildSearch()', () => {
    it('correctly builds the Set', () => {
      const index1 = {
        foo: 1,
        bar: 2,
        bazz: 3,
        fizz: 4,
      };

      const index2 = {
        foo: 3,
        bar: 4,
        bazz: 5,
        fizz: 6,
      };

      const excludeList = [3];

      const search = utils.buildSearch([index1, index2], [excludeList]);

      expect(search).toEqual(new Set([1, 2, 4, 5, 6]));
    });
  });

  describe('getVariableIndex()', () => {
    it('extracts variables into index', () => {
      const subject = getVariableIndex(testState);

      expect(subject).toMatchSnapshot();
    });
  });

  describe('getAssetIndex()', () => {
    it('extracts asset references into index', () => {
      const subject = getAssetIndex(testState);

      expect(subject).toMatchSnapshot();
    });
  });

  describe('getNodeIndex()', () => {
    it('extracts subject references into type index', () => {
      const subject = getNodeIndex(testState);

      expect(subject).toMatchSnapshot();
    });
  });

  describe('getEdgeIndex()', () => {
    it('extracts subject references into type index', () => {
      const subject = getEdgeIndex(testState);

      expect(subject).toMatchSnapshot();
    });
  });
});
