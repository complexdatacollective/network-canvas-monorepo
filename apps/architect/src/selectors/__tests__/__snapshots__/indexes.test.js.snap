// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`indexes selectors getAssetIndex() extracts asset references into index 1`] = `
Object {
  "stages[12].background.image": "quadrant-1",
  "stages[13].dataSource": "previousInterview-json",
  "stages[14].dataSource": "HIVServices",
  "stages[1].panels[0].dataSource": "existing",
  "stages[1].panels[1].dataSource": "previousInterview-json",
  "stages[20].items[0].content": "click_the_thing",
  "stages[20].items[1].content": "rubberduck",
  "stages[20].items[2].content": "rubberduck",
  "stages[21].items[1].content": "withSound",
  "stages[2].panels[0].dataSource": "previousInterview-csv",
  "stages[3].panels[0].dataSource": "existing",
  "stages[8].dataSource": "HIVServices-csv",
  "stages[9].dataSource": "previousInterview-json",
}
`;

exports[`indexes selectors getEdgeIndex() extracts subject references into type index 1`] = `
Object {
  "stages[11].prompts[0].edges.create": "77199445-9d50-4646-b0bc-6d6b0c0e06bd",
  "stages[11].prompts[0].edges.display[0]": "867127d9-086b-403a-a3e7-2c7d32126546",
  "stages[11].prompts[0].edges.display[1]": "77199445-9d50-4646-b0bc-6d6b0c0e06bd",
  "stages[11].prompts[1].edges.create": "867127d9-086b-403a-a3e7-2c7d32126546",
  "stages[11].prompts[1].edges.display[0]": "77199445-9d50-4646-b0bc-6d6b0c0e06bd",
  "stages[11].prompts[1].edges.display[1]": "867127d9-086b-403a-a3e7-2c7d32126546",
  "stages[11].prompts[2].edges.display[0]": "77199445-9d50-4646-b0bc-6d6b0c0e06bd",
  "stages[11].prompts[2].edges.display[1]": "867127d9-086b-403a-a3e7-2c7d32126546",
  "stages[11].prompts[3].edges.create": "77199445-9d50-4646-b0bc-6d6b0c0e06bd",
  "stages[11].prompts[3].edges.display[0]": "77199445-9d50-4646-b0bc-6d6b0c0e06bd",
  "stages[11].prompts[3].edges.display[1]": "867127d9-086b-403a-a3e7-2c7d32126546",
  "stages[17].subject.type": "77199445-9d50-4646-b0bc-6d6b0c0e06bd",
  "stages[18].subject.type": "867127d9-086b-403a-a3e7-2c7d32126546",
  "stages[19].presets[0].edges.display[0]": "77199445-9d50-4646-b0bc-6d6b0c0e06bd",
  "stages[19].presets[0].edges.display[1]": "867127d9-086b-403a-a3e7-2c7d32126546",
  "stages[19].presets[1].edges.display[0]": "867127d9-086b-403a-a3e7-2c7d32126546",
  "stages[4].prompts[0].createEdge": "77199445-9d50-4646-b0bc-6d6b0c0e06bd",
  "stages[4].prompts[1].createEdge": "867127d9-086b-403a-a3e7-2c7d32126546",
  "stages[5].prompts[0].createEdge": "77199445-9d50-4646-b0bc-6d6b0c0e06bd",
  "stages[5].prompts[1].createEdge": "867127d9-086b-403a-a3e7-2c7d32126546",
}
`;

exports[`indexes selectors getNodeIndex() extracts subject references into type index 1`] = `
Object {
  "stages[10].subject.type": "c93af3fe-9f84-4e85-acb5-5b15f30788c7",
  "stages[11].subject.type": "4aebf73e-95e3-4fd1-95e7-237dcc4a4466",
  "stages[12].subject.type": "eda5e3bb-8e1c-4216-9e06-adc0ff6b7f73",
  "stages[13].subject.type": "4aebf73e-95e3-4fd1-95e7-237dcc4a4466",
  "stages[14].subject.type": "eda5e3bb-8e1c-4216-9e06-adc0ff6b7f73",
  "stages[15].subject.type": "4aebf73e-95e3-4fd1-95e7-237dcc4a4466",
  "stages[16].subject.type": "4aebf73e-95e3-4fd1-95e7-237dcc4a4466",
  "stages[19].subject.type": "4aebf73e-95e3-4fd1-95e7-237dcc4a4466",
  "stages[1].subject.type": "4aebf73e-95e3-4fd1-95e7-237dcc4a4466",
  "stages[2].subject.type": "4aebf73e-95e3-4fd1-95e7-237dcc4a4466",
  "stages[3].subject.type": "eda5e3bb-8e1c-4216-9e06-adc0ff6b7f73",
  "stages[4].subject.type": "4aebf73e-95e3-4fd1-95e7-237dcc4a4466",
  "stages[5].subject.type": "4aebf73e-95e3-4fd1-95e7-237dcc4a4466",
  "stages[6].subject.type": "4aebf73e-95e3-4fd1-95e7-237dcc4a4466",
  "stages[7].subject.type": "eda5e3bb-8e1c-4216-9e06-adc0ff6b7f73",
  "stages[8].subject.type": "eda5e3bb-8e1c-4216-9e06-adc0ff6b7f73",
  "stages[9].subject.type": "4aebf73e-95e3-4fd1-95e7-237dcc4a4466",
}
`;

exports[`indexes selectors getVariableIndex() extracts variables into index 1`] = `
Object {
  "codebook.edge[77199445-9d50-4646-b0bc-6d6b0c0e06bd].variables[e343a91f-628d-4175-870c-957beffa0003].validation.sameAs": "e343a91f-628d-4175-870c-957beffa0002",
  "codebook.ego.variables[9c47f32e-b72f-4b1b-93a2-e29766dc3012].validation.differentFrom": "23a10bcb-bd71-46a1-a36d-4ab80cb5a1d1",
  "codebook.node[4aebf73e-95e3-4fd1-95e7-237dcc4a4466].variables[0e75ec18-2cb1-4606-9f18-034d28b07c19].validation.differentFrom": "6be95f85-c2d9-4daf-9de1-3939418af888",
  "stages[0].form.fields[0].variable": "3377af3f-3c79-41da-9b0b-6570fb519b93",
  "stages[0].form.fields[10].variable": "9c47f32e-b72f-4b1b-93a2-e29766dc3012",
  "stages[0].form.fields[1].variable": "b4956387-cada-412f-9e97-fe352c3cc44e",
  "stages[0].form.fields[2].variable": "f3eb5f0b-84ad-43f0-8db5-9f83cd3a56de",
  "stages[0].form.fields[3].variable": "683ad529-66c6-4c84-8e31-a31452d6b5e9",
  "stages[0].form.fields[4].variable": "f3537457-41a2-4fbf-a1f8-9235b3c1edc6",
  "stages[0].form.fields[5].variable": "7c647977-f2d1-426f-b0c5-6d44c795c554",
  "stages[0].form.fields[6].variable": "67fefac9-561f-454a-8f78-81bd9470efaa",
  "stages[0].form.fields[7].variable": "300a798f-d11b-4195-91ee-3221bc586c0a",
  "stages[0].form.fields[8].variable": "03c54247-aa12-497a-b8f8-c4cb69048455",
  "stages[0].form.fields[9].variable": "23a10bcb-bd71-46a1-a36d-4ab80cb5a1d1",
  "stages[10].form.fields[0].variable": "2377af3f-3c79-41da-9b0b-6570fb519b93",
  "stages[10].form.fields[1].variable": "e3eb5f0b-84ad-43f0-8db5-9f83cd3a56de",
  "stages[10].form.fields[2].variable": "583ad529-66c6-4c84-8e31-a31452d6b5e9",
  "stages[10].form.fields[3].variable": "e3537457-41a2-4fbf-a1f8-9235b3c1edc6",
  "stages[10].form.fields[4].variable": "6c647977-f2d1-426f-b0c5-6d44c795c554",
  "stages[10].form.fields[5].variable": "57fefac9-561f-454a-8f78-81bd9470efaa",
  "stages[10].form.fields[6].variable": "683ad529-66c6-4c84-8e31-a31452dsf74k",
  "stages[10].form.fields[7].variable": "e4e0df51-4b53-4e16-9249-8587537370e9",
  "stages[10].form.fields[8].variable": "720224a3-e2d3-4729-968a-29d3f3e4fca6",
  "stages[11].filter.rules[0].options.attribute": "6be95f85-c2d9-4daf-9de1-3939418af888",
  "stages[11].filter.rules[1].options.attribute": "6be95f85-c2d9-4daf-9de1-3939418af888",
  "stages[11].prompts[0].layout.layoutVariable": "d13ca72d-aefe-4f48-841d-09f020e0e988",
  "stages[11].prompts[1].layout.layoutVariable": "d13ca72d-aefe-4f48-841d-09f020e0e988",
  "stages[11].prompts[2].highlight.variable": "03b03617-46ae-41cb-9462-9acd8a17edd6",
  "stages[11].prompts[2].layout.layoutVariable": "d13ca72d-aefe-4f48-841d-09f020e0e988",
  "stages[11].prompts[3].highlight.variable": "03b03617-46ae-41cb-9462-9acd8a17edd6",
  "stages[11].prompts[3].layout.layoutVariable": "d13ca72d-aefe-4f48-841d-09f020e0e988",
  "stages[12].prompts[0].layout.layoutVariable": "0a5f7952-a9a7-4e87-9353-2d17a592ssss",
  "stages[13].cardOptions.additionalProperties[0].variable": "last_name",
  "stages[13].cardOptions.additionalProperties[1].variable": "age",
  "stages[13].prompts[0].additionalAttributes[0].variable": "ae44b6ba-ad49-4ecf-bf38-b65decfce000",
  "stages[13].prompts[1].additionalAttributes[0].variable": "e1dda563-f6f6-47b7-91cb-31b4c2fdb73a",
  "stages[13].searchOptions.matchProperties[0]": "variableNotInRegistry",
  "stages[13].searchOptions.matchProperties[1]": "nickname",
  "stages[13].searchOptions.matchProperties[2]": "last_name",
  "stages[14].cardOptions.additionalProperties[0].variable": "67132d2b-c371-4c57-a5eb-6520083f9d22",
  "stages[14].cardOptions.additionalProperties[1].variable": "8ee3a187-d4be-458e-8abb-71efcc071949",
  "stages[14].prompts[0].additionalAttributes[0].variable": "55f1fbbe-2fe5-42a7-88fb-9a8e5e659d2f",
  "stages[14].searchOptions.matchProperties[0]": "1e9ce62c-44c5-484d-9c26-0d20cc7d7238",
  "stages[14].searchOptions.matchProperties[1]": "8ee3a187-d4be-458e-8abb-71efcc071949",
  "stages[14].searchOptions.matchProperties[2]": "67132d2b-c371-4c57-a5eb-6520083f9d22",
  "stages[15].filter.rules[0].options.attribute": "6be95f85-c2d9-4daf-9de1-3939418af888",
  "stages[15].prompts[0].variable": "e343a91f-628d-4175-870c-957beffa0152",
  "stages[15].prompts[1].otherVariable": "979c9ad7-74cb-402d-8574-3f0a91c9174d",
  "stages[15].prompts[1].variable": "e343a91f-628d-4175-870c-957beffa0153",
  "stages[15].prompts[2].variable": "e343a91f-628d-4175-870c-957beffa0154",
  "stages[15].prompts[3].binSortOrder[0].property": "0e75ec18-2cb1-4606-9f18-034d28b07c19",
  "stages[15].prompts[3].binSortOrder[1].property": "c5fee926-855d-4419-b5bb-54e89010cea6",
  "stages[15].prompts[3].binSortOrder[2].property": "0ff25001-a2b8-46de-82a9-53143aa00d10",
  "stages[15].prompts[3].bucketSortOrder[0].property": "0e75ec18-2cb1-4606-9f18-034d28b07c19",
  "stages[15].prompts[3].variable": "e343a91f-628d-4175-870c-957beffa0151",
  "stages[15].prompts[4].variable": "4ccba4e2-a246-46ee-a3ff-1c70fb4c449c",
  "stages[16].form.fields[0].variable": "246c4d84-ebd7-41a6-8c1a-ef83938227ba",
  "stages[16].form.fields[1].variable": "9b56bc6d-6b20-4e38-9aab-4057e78e1130",
  "stages[16].form.fields[2].variable": "23c5a7b9-b553-4ff1-b515-6a81999773d2",
  "stages[16].form.fields[3].variable": "ae44b6ba-ad49-4ecf-bf38-b65decfce000",
  "stages[16].form.fields[4].variable": "4ccba4e2-a246-46ee-a3ff-1c70fb4c449c",
  "stages[16].form.fields[5].variable": "490db003-3289-4969-b426-919cefb4cdf0",
  "stages[16].skipLogic.filter.rules[0].options.attribute": "e343a91f-628d-4175-870c-957beffa0153",
  "stages[17].form.fields[0].variable": "e343a91f-628d-4175-870c-957beffa0002",
  "stages[17].form.fields[1].variable": "e343a91f-628d-4175-870c-957beffa0003",
  "stages[17].form.fields[2].variable": "e343a91f-628d-4175-870c-957beffa0004",
  "stages[18].form.fields[0].variable": "d343a91f-628d-4175-870c-957beffa0002",
  "stages[19].presets[0].edges.display[0]": "77199445-9d50-4646-b0bc-6d6b0c0e06bd",
  "stages[19].presets[0].edges.display[1]": "867127d9-086b-403a-a3e7-2c7d32126546",
  "stages[19].presets[0].groupVariable": "e343a91f-628d-4175-870c-957beffa0154",
  "stages[19].presets[0].highlight[0]": "03b03617-46ae-41cb-9462-9acd8a17edd6",
  "stages[19].presets[0].highlight[1]": "ae44b6ba-ad49-4ecf-bf38-b65decfce000",
  "stages[19].presets[0].layoutVariable": "d13ca72d-aefe-4f48-841d-09f020e0e988",
  "stages[19].presets[1].edges.display[0]": "867127d9-086b-403a-a3e7-2c7d32126546",
  "stages[19].presets[1].groupVariable": "e343a91f-628d-4175-870c-957beffa0154",
  "stages[19].presets[1].highlight[0]": "03b03617-46ae-41cb-9462-9acd8a17edd6",
  "stages[19].presets[1].layoutVariable": "e13ca72d-aefe-4f48-841d-09f020e0e987",
  "stages[1].form.fields[0].variable": "6be95f85-c2d9-4daf-9de1-3939418af888",
  "stages[1].form.fields[1].variable": "0e75ec18-2cb1-4606-9f18-034d28b07c19",
  "stages[1].form.fields[2].variable": "c5fee926-855d-4419-b5bb-54e89010cea6",
  "stages[1].form.fields[3].variable": "e343a91f-628d-4175-870c-957beffa0154",
  "stages[1].prompts[0].additionalAttributes[0].variable": "ae44b6ba-ad49-4ecf-bf38-b65decfce000",
  "stages[1].prompts[0].additionalAttributes[1].variable": "94e2dee9-1c25-4b64-b8bb-4946465c3c07",
  "stages[1].prompts[1].additionalAttributes[0].variable": "94e2dee9-1c25-4b64-b8bb-4946465c3c07",
  "stages[23].skipLogic.filter.rules[0].options.attribute": "6be95f85-c2d9-4daf-9de1-3939418af888",
  "stages[23].skipLogic.filter.rules[1].options.attribute": "3377af3f-3c79-41da-9b0b-6570fb519b93",
  "stages[2].form.fields[0].variable": "6be95f85-c2d9-4daf-9de1-3939418af888",
  "stages[2].form.fields[1].variable": "0e75ec18-2cb1-4606-9f18-034d28b07c19",
  "stages[2].form.fields[2].variable": "c5fee926-855d-4419-b5bb-54e89010cea6",
  "stages[3].prompts[0].additionalAttributes[0].variable": "55f1fbbe-2fe5-42a7-88fb-9a8e5e659d2f",
  "stages[3].quickAdd": "66646d68-cd33-407b-a349-56707021df72",
  "stages[5].prompts[0].edgeVariable": "e343a91f-628d-4175-870c-75429857399e",
  "stages[5].prompts[1].edgeVariable": "f984jk23-628d-4175-870c-ejd8e7al8vc3",
  "stages[6].filter.rules[0].options.attribute": "ae44b6ba-ad49-4ecf-bf38-b65decfce000",
  "stages[6].prompts[0].binSortOrder[0].property": "0e75ec18-2cb1-4606-9f18-034d28b07c19",
  "stages[6].prompts[0].binSortOrder[1].property": "c5fee926-855d-4419-b5bb-54e89010cea6",
  "stages[6].prompts[0].bucketSortOrder[0].property": "0e75ec18-2cb1-4606-9f18-034d28b07c19",
  "stages[6].prompts[0].variable": "1096204b-48fe-444c-b642-4ab211f7f57c",
  "stages[6].prompts[1].variable": "23c5a7b9-b553-4ff1-b515-6a81999773d2",
  "stages[7].prompts[0].variable": "a39a86ef-18c7-4545-8fbc-d0e5894565d0",
  "stages[9].cardOptions.additionalProperties[0].variable": "last_name",
  "stages[9].cardOptions.additionalProperties[1].variable": "age",
  "stages[9].prompts[0].additionalAttributes[0].variable": "ae44b6ba-ad49-4ecf-bf38-b65decfce000",
}
`;
